crypto/multisig/src/v2.rs

```diff
@@ -0,0 +1,1259 @@
+    /// threshold of both groups but since A gets "consumed" when matching the
+    /// first group, leaving the second group with only D and E, it will not
+    /// be considered a match. We say we could match all 5 signatures
+    /// in theory, because if we ignored A for the first set, it could still be
+    /// used to match the 2nd set. Such algorithm would require a more
+    /// complex (and slower) implementation, and is not currently supported.
+    pub fn verify<
+        SIG: Clone
+            + Default
+            + Digestible
+            + Eq
+            + Hash
+            + Message
+            + Ord
+            + PartialEq
+            + PartialOrd
+            + Serialize
+            + Signature,
+    >(
+        &self,
+        message: &[u8],
+        multi_sig: &MultiSig<SIG>,
+    ) -> Result<Vec<S>, SignatureError>
+    where
+        S: Verifier<SIG>,
+    {
+        if multi_sig.signatures().len() > MAX_SIGNATURES {
+            return Err(SignatureError::new());
+        }
+
+        // Sort and dedup the list of signatures..
+        let mut signatures = multi_sig.signatures.clone();
+        signatures.sort_by(|a, b| a.as_ref().cmp(b.as_ref()));
+        signatures.dedup();
+
+        // Verify signatures.
+        self.verify_helper(message, &signatures, &[])
+    }
+
+    fn verify_helper<
+        SIG: Clone
+            + Default
+            + Digestible
+            + Eq
+            + Hash
+            + Message
+            + Ord
+            + PartialEq
+            + PartialOrd
+            + Serialize
+            + Signature,
+    >(
+        &self,
+        message: &[u8],
+        signatures: &[SIG],
+        seen_signers: &[S],
+    ) -> Result<Vec<S>, SignatureError>
+    where
+        S: Verifier<SIG>,
+    {
+        // Sort and dedup the list of signers.
+        // While the verification code below should be immune to duplicate signers, the
+        // overhead of deduping them is negligible and being extra-safe is a good idea.
+        let mut signers = self.signers.clone();
+        signers.sort();
+        signers.dedup();
+
+        // See which signers are satisfied by which signatures.
+        let mut matched_signers = Vec::new();
+        let mut num_matched_entities = 0;
+
+        for signer in signers.iter() {
+            match signer.entity {
+                Some(SignerEntity::Single(ref s)) => {
+                    // If we already encountered this signer, we cannot use it again.
+                    if seen_signers.contains(s) {
+                        continue;
+                    }
+                    if matched_signers.contains(s) {
+                        continue;
+                    }
+
+                    for signature in signatures.iter() {
+                        if s.verify(message, signature).is_ok() {
+                            matched_signers.push(s.clone());
+                            num_matched_entities += 1;
```

Comment:
continuing from earlier discussion:

https://github.com/mobilecoinfoundation/mobilecoin/pull/2657/files#r990539556

> Pretty sure we still want code that makes sure you are not handed a bunch of differently-looking signatures that all match a single signer at the level of a single set (which is what the current implementation does)

In this code I'm reading, 

```
                    for signature in signatures.iter() {
                        if s.verify(message, signature).is_ok() {
                            matched_signers.push(s.clone());
                            num_matched_entities += 1;
                            break;
                        }
                    }
```

I would suggest it would be okay to rewrite this like

```
if signatures.iter().filter(|signature| s.verify(message, signature).is_ok()).any() {
   num_matched_entities += 1;
}
```

I think it would also be okay to do a stricter but slower check here like:

```
match signatures.iter().filter(|signature| s.verify(message, signature).is_ok()).count() {
   0 => {}
   1 => { num_matched_entities += 1; }
   n => { return Err(Error::SquirrelySignatureMatchedMultipleSignerIdentities(n)); }
}
```

But I just don't think it's super interesting to maintain the `matched_signers` here because:

1. We call `signers.dedup()` anyways so the check `if matched_signers.contains(s)` should always be false.
2. My point of view is that the threat model for this code is that, the Signers are the good guys, and they are trusted. The adversary is potentially creating bad signatures and trying to get them verified. But if the signers are generating their identities randomly, then it isn't going to be possible to make a signature that satisfies two different signer identities -- they aren't going to be related by small cofactor or anything like that if they are generated securely.
3. It seems possible to me that the signers would want to allow for one signing identity to count for two votes. For example, when Henry and I set up the testnet signing stuff, we set up a bunch of Ledger hardware wallet based identities, but then we also created some "backup" identities that were not based on the Ledger, so that if somehow the ledger hardware or software gets screwed up, we don't get locked out. However, the higher the threshold you want to set on the signature set, the more backups you need to overcome things like that. If you could just repeat the failsafe backup several times in the signer set, then you don't have that issue.

I am not sure if (3) is actually an important consideration for practitioners in production, and it's not, then it doesn't really matter. But like, philosophically, because my view of the threat model is that the signer set is produced by the good guys, I don't think it's bad if one signature counts multiple times within a signer set. I think that's at least plausibly useful, and I think it can only happen if the good guys explicitly want it to happen, so we should allow it to happen in the code.

Reply:
IIRC one of the main concerns in all of this was that, given one valid ed25519 signature, you can perturb it in ways so that the bytes are different but it is valid against the same keys as before. That makes it harder to prevent replay attacks "just" by comparing the bytes of signature sets. But we avoided this concern in a different way by using the nonce stuff when minting.

I think as long as we document our multisignature stuff in a way that mentions that this kind of replay issue needs to be prevented by the user of this crate, then I'm not sure what issues actually remain about signature / signer multiplicity.

Reply:
I feel like this is similar to what others are saying here and here:

https://crypto.stackexchange.com/questions/96143/should-ed25519-verification-multiply-by-the-cofactor
https://crypto.stackexchange.com/questions/55632/libsodium-x25519-and-ed25519-small-order-check/55643#55643

> In a signature system, you rely on verification under known public keys; a malicious signer giving one of the ‘blacklisted’ points is out of scope. Blacklisted points in signatures can't fool verifiers—the signature scheme would be horribly broken if they could.
>
> For more details on validation, see [Trevor Perrin's article on point validation in such protocols](https://moderncrypto.org/mail-archive/curves/2017/000896.html).

Trevor gives an example I feel is related:

>  * The essay claims a "non-obvious attack" against Signal where Alice
chooses insecure keys for herself so that messages encrypted to her
use insecure keys.  Of course, it's absurd to consider Alice attacking
herself.

If we're going to say the threat model is that, signers might choose degenerate / related keys in order to undermine the signature scheme, allowing attackers to create signatures that match to multiple signer identities in a way that Signers didn't expect, then we are considering a model in which the Signers are attacking themselves.

Reply:
> Pretty sure we still want code that makes sure you are not handed a bunch of differently-looking signatures that all match a single signer at the level of a single set (which is what the current implementation does)

I know that when we first discussed this, we decided that this is a goal for the current implementation. What I'm saying is that I think there was actually no good reason for this, it just created complexity, which now gets even more complex when we try to extend the feature. I think what happened was, we were fixating on all the different forms of subtle malleability in ed25519 without thinking enough about what actually matters in the threat model where we are trying to use it.